base/compiler/optimize.jl

```diff
@@ -628,7 +628,8 @@ function is_pure_intrinsic_infer(f::IntrinsicFunction)
 end
 
 # whether `f` is effect free if nothrow
-intrinsic_effect_free_if_nothrow(f) = f === Intrinsics.pointerref || is_pure_intrinsic_infer(f)
+intrinsic_effect_free_if_nothrow(f) = f === Intrinsics.pointerref ||
+    f === Intrinsics.have_fma || is_pure_intrinsic_infer(f)
```

Comment:
```suggestion
intrinsic_effect_free_if_nothrow(f) = 
    f === Intrinsics.pointerref ||
    # in order to allow FMA multiversioning, we exclude `have_fma` from `is_pure_intrinsic_infer` 
    # because it shouldn't be constant folded at inference time, but we still mark it here 
    # so that we allow a call with `have_fma` to use the constant calling convention 
    # when both branches have the same constant value.
    f === Intrinsics.have_fma ||
    is_pure_intrinsic_infer(f)
```

Reply:
Multiversioning is one of the issues, but it's not exclusive. Everything here that reads implicit runtime state needs to be excluded. If it is not guaranteed to always return bitwise egal output given bitwise egal input, it must be excluded from `is_pure_intrinsic_infer`. I'm working on a PR to make the effect free code a little more principled, so I'll fold the documentation comment in there.